pkg/ingester/circuitbreaker.go

```diff
@@ -0,0 +1,227 @@
+		}).
+		HandleIf(func(_ any, err error) bool { return isFailure(err) })
+
+	if cfg.testModeEnabled {
+		// In case of testing purposes, we initialize the circuit breaker with count based failure thresholding,
+		// since it is more deterministic, and therefore it is easier to predict the outcome.
+		cbBuilder = cbBuilder.WithFailureThreshold(cfg.FailureThreshold)
+	} else {
+		// In case of production code, we prefer time based failure thresholding.
+		cbBuilder = cbBuilder.WithFailureRateThreshold(cfg.FailureThreshold, cfg.FailureExecutionThreshold, cfg.ThresholdingPeriod)
+	}
+
+	cb := cbBuilder.Build()
+	return &circuitBreaker{
+		CircuitBreaker: cb,
+		ingester:       ingester,
+		executor:       failsafe.NewExecutor[any](cb),
+		startTime:      time.Now().Add(cfg.InitialDelay),
+	}
+}
+
+func isFailure(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// We only consider timeouts or ingester hitting a per-instance limit
+	// to be errors worthy of tripping the circuit breaker since these
+	// are specific to a particular ingester, not a user or request.
+
+	if errors.Is(err, context.DeadlineExceeded) {
+		return true
+	}
+
+	var ingesterErr ingesterError
+	if errors.As(err, &ingesterErr) {
+		return ingesterErr.errorCause() == mimirpb.INSTANCE_LIMIT
+	}
+
+	return false
+}
+
+func (cb *circuitBreaker) isActive() bool {
+	return cb.startTime.Before(time.Now())
+}
+
+func (cb *circuitBreaker) ingesterID() string {
+	return cb.ingester.cfg.IngesterRing.InstanceID
+}
+
+func (cb *circuitBreaker) logger() log.Logger {
+	return cb.ingester.logger
+}
+
+func (cb *circuitBreaker) metrics() *ingesterMetrics {
+	return cb.ingester.metrics
+}
+
+func (cb *circuitBreaker) config() CircuitBreakerConfig {
+	return cb.ingester.cfg.CircuitBreakerConfig
+}
+
+func (cb *circuitBreaker) get(ctx context.Context, callbackFnName string, callbackFn func() (any, error)) (any, error) {
+	res, err := cb.executor.Get(callbackFn)
+	if err != nil && errors.Is(err, circuitbreaker.ErrOpen) {
+		cb.metrics().circuitBreakerResults.WithLabelValues(cb.ingesterID(), resultOpen).Inc()
+		cbOpenErr := middleware.DoNotLogError{Err: newCircuitBreakerOpenError(cb.RemainingDelay())}
+		return res, newErrorWithStatus(cbOpenErr, codes.Unavailable)
+	}
+	return res, cb.processError(ctx, err, callbackFnName)
+}
+
+func (cb *circuitBreaker) processError(ctx context.Context, err error, callbackFnName string) error {
+	if err == nil {
+		return nil
+	}
+	if errors.Is(err, ctx.Err()) {
+		level.Error(cb.logger()).Log("msg", fmt.Sprintf("callback function %s completed with an error found in the context", callbackFnName), "ingester", cb.ingesterID(), "ctxErr", ctx.Err())
+
+		// ctx.Err() was registered with the circuit breaker's executor, but we don't propagate it
+		return nil
+	}
+
+	level.Error(cb.logger()).Log("msg", fmt.Sprintf("callback function %s completed with an error", callbackFnName), "ingester", cb.ingesterID(), "err", err)
```

Comment:
Do we want to log these errors? Eg. for Push errors, they are already logged by logging middleware (and possibly sampled), and `StartPushRequest` we don't want to log those.

I think we can drop this logging altogether.